asyncpg/protocol/coreproto.pyx

```diff
@@ -4,8 +4,212 @@
+            raise Exception("could not get iterations")
+
+    cdef verify_server_final_message(self, bytes server_final_message):
+        """Verify the final message from the server"""
+        cdef:
+            bytes server_signature
+
+        try:
+            server_signature = re.search(b'v=([^,]+)', server_final_message).group(1)
+        except IndexError:
+            raise Exception("could not get server signature")
+
+        verify_server_signature = hmac.new(self.server_key.digest(), self.authorization_message, self.DIGEST)
+        # validate the server signature against the verifier
+        return server_signature == base64.b64encode(verify_server_signature.digest())
+
+    cdef _bytes_xor(self, bytes a, bytes b):
+        """XOR two bytestrings together"""
+        return bytes(a_i ^ b_i for a_i, b_i in zip(a, b))
+
+    cdef _generate_client_nonce(self, int num_bytes):
+        cdef:
+            bytes token
+
+        token = secrets.token_bytes(num_bytes) if secrets else os.urandom(num_bytes)
+
+        return base64.b64encode(token)
 
-from hashlib import md5 as hashlib_md5  # for MD5 authentication
+    cdef _generate_client_proof(self, str password):
+        """need to ensure a server response exists, i.e. """
+        cdef:
+            bytes salted_password
+
+        if any([getattr(self, val) is None for val in self.REQUIREMENTS_CLIENT_PROOF]):
+            raise Exception("you need values from server in order to generate a client proof")
+        # generate a salt password
+        salted_password = self._generate_salted_password(password, self.password_salt, self.password_iterations)
+        # client key is derived from the salted password
+        client_key = hmac.new(salted_password, b"Client Key", self.DIGEST)
+        # this allows us to compute the stored key that is residing on the server
+        stored_key = self.DIGEST(client_key.digest())
+        # as well as compute the server key
+        self.server_key = hmac.new(salted_password, b"Server Key", self.DIGEST)
+        # build the authorization message that will be used in the client signature
+        # the "c=" portion is for the channel binding, but this is not presently set
+        self.authorization_message = self.client_first_message_bare + b"," + self.server_first_message + \
+            b",c=" + base64.b64encode(self.client_channel_binding) + b",r=" +  self.server_nonce
+        # sign!
+        client_signature = hmac.new(stored_key.digest(), self.authorization_message, self.DIGEST)
+        # and the proof
+        return self._bytes_xor(client_key.digest(), client_signature.digest())
+
+    cdef _generate_salted_password(self, str password, bytes salt, int iterations):
+        """This follows the "Hi" algorithm specified in RFC5802"""
+        cdef:
+            bytes p
+            bytes s
+            bytes u
+
+        # convert the password to a binary string - UTF8 is safe for SASL (though there are SASLPrep rules)
```

Comment:
SASLprep is not optional for SCRAM, it must be implemented, or certain passwords will fail.  Here is a reasonable implementation: https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/saslprep.py

Reply:
To clarify: per Postgres docs, passwords that fail SASLprep should be sent as-is, but the normalization rules still apply.